### PR TITLE
Improve API Godocs + add API unit tests

### PR DESCRIPTION
**Pull request by hugelgupf**
cc @bradfitz just as FYI for looking at the resulting API if you want.

Before:

https://godoc.org/github.com/u-root/u-root/pkg/uroot

After:

https://godoc.org/gopkg.in/hugelgupf/u-root.v10/pkg/uroot
https://godoc.org/gopkg.in/hugelgupf/u-root.v10/pkg/uroot/builder
https://godoc.org/gopkg.in/hugelgupf/u-root.v10/pkg/uroot/initramfs

There is *one* non-code-moving change here up to the pkg/ls commit, which is to change `type Builder` to an interface.

### CODE REVIEW

**Comment by hugelgupf** [comment]
I'll probably split this back into two PRs - one for docs, one for added unit tests.

File: pkg/ls/fileinfo_unix.go [hunk 1]
```diff
@@ -91,60 +96,61 @@ func lookupGroupName(id uint32) string {
 	return s
 }
 
-// The default stringer. Return only the filename. Unprintable characters are
-// replaced with '?'.
-func (fi fileInfo) String() string {
-	return unprintableRe.ReplaceAllLiteralString(fi.name, "?")
+type Stringer interface {
```
Comment (rjoleary) [inline pkg/ls/fileinfo_unix.go]: Needs godoc string because it is exported.

File: pkg/ls/fileinfo_unix.go [hunk 2]
```diff
@@ -91,60 +96,61 @@ func lookupGroupName(id uint32) string {
...
+	FileString(fi FileInfo) string
 }
 
-// Two alternative stringers
-type quotedStringer struct {
-	fileInfo
+type NameStringer struct{}
```
Comment (rjoleary) [inline pkg/ls/fileinfo_unix.go]: Needs godoc string because it is exported.

File: pkg/ls/fileinfo_unix.go [hunk 3]
```diff
@@ -91,60 +96,61 @@ func lookupGroupName(id uint32) string {
...
+
+func (ns NameStringer) FileString(fi FileInfo) string {
```
Comment (rjoleary) [inline pkg/ls/fileinfo_unix.go]: Needs godoc string because it is exported.

File: pkg/ls/fileinfo_unix.go [hunk 4]
```diff
@@ -91,60 +96,61 @@ func lookupGroupName(id uint32) string {
...
+
+// FileString returns the name surrounded by quotes with escaped control characters.
+func (qs QuotedStringer) FileString(fi FileInfo) string {
+	return fmt.Sprintf("%#v", fi.Name)
 }
 
-// Return the name surrounded by quotes with escaped control characters.
-func (fi quotedStringer) String() string {
-	return fmt.Sprintf("%#v", fi.name)
+type LongStringer struct {
```
Comment (rjoleary) [inline pkg/ls/fileinfo_unix.go]: Needs godoc string because it is exported.

File: pkg/ls/fileinfo_unix.go [hunk 5]
```diff
@@ -91,60 +96,61 @@ func lookupGroupName(id uint32) string {
...
+	Human bool
+	Name  Stringer
 }
 
-// The long and quoted stringers can be combined like so:
-//     longStringer{fi, quotedStringer{fi}}
-func (fi longStringer) String() string {
+func (ls LongStringer) FileString(fi FileInfo) string {
```
Comment (rjoleary) [inline pkg/ls/fileinfo_unix.go]: Needs godoc string because it is exported.

File: pkg/cpio/newc_test.go [hunk 6]
```diff
@@ -15,6 +15,8 @@ import (
 	"github.com/u-root/u-root/pkg/uio"
 )
 
+/*
+This test should really compare against []Records, not the string formatted []Records.
```
Comment (rjoleary) [inline pkg/cpio/newc_test.go]: Put TODO here, so it's easier to track.

File: pkg/uroot/builder/bb/bb.go [hunk 7]
```diff
@@ -49,6 +72,14 @@ func BuildBusybox(env golang.Environ, pkgs []string, binaryPath string) error {
 		return err
 	}
 
+	bblock := filepath.Join(urootPkg.Dir, "bblock")
+	// Only one busybox can be compiled at a time.
```
Comment (rjoleary) [inline pkg/uroot/builder/bb/bb.go]: Would be useful to have an explanation for why.

File: pkg/uroot/uroot.go [hunk 8]
```diff
@@ -64,63 +51,101 @@ var DefaultRamfs = []cpio.Record{
...
+//
+// See Builder for an explanation of build modes.
 type Commands struct {
-	// Builder is the build format.
-	Builder Builder
+	// Builder is the Go compiler mode.
+	Builder builder.Builder
 
-	// Packages are the Go packages to add to the archive.
+	// Packages are the Go packages to compile and add to the archive.
```
[the remark this answers is not in the document]
Comment (hugelgupf) [reply pkg/uroot/uroot.go]: No, but I'm not sure how to generalize that in a way that is still easy to read.

File: pkg/uroot/uroot.go [hunk 9]
```diff
@@ -221,14 +323,16 @@ func ResolvePackagePaths(env golang.Environ, pkgs []string) ([]string, error) {
...
-// parsed from the following formats:
+// ParseExtraFiles adds files from the extraFiles list to the archive.
+//
+// The following formats are allowed in the extraFiles list:
 //
-// - hostPath:archivePath adds the file from hostPath at the relative archivePath in the archive.
-// - justAPath is added to the archive under justAPath.
+//   - `hostPath:archivePath` adds the file from hostPath at the relative
+//     archivePath in the archive.
+//   - `justAPath` is added to the archive under justAPath.
```
Comment (rjoleary) [inline pkg/uroot/uroot.go]: Do the same to this `justAPath`.

File: pkg/uroot/uroot.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (rjoleary) [reply pkg/uroot/uroot.go]: How about: "Packages lists Go packages to be included (complied or otherwise) in the archive"?

**Pull request merged**